### PR TITLE
cluster up - sane image defaults

### PR DESCRIPTION
**Pull request by csrwng**
Proposed fix for https://github.com/openshift/origin/issues/9848

Sets the default value for the --version flag to the release that corresponds to the `oc` client. This matches the behavior for `openshift start`.

Introduces a new flag `--always-pull` to always pull an image before using it. 

Therefore, the default is not to use whatever :latest is on the current machine, but the last released set of images.

If you want to use whatever is on your machine, you can always specify `--version=latest`.

If you want the absolute set of latest images (daily builds), you can specify `--version=latest --always-pull`

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees @tnozicka please let me know what you think

File: pkg/bootstrap/docker/up.go [hunk 1]
```diff
@@ -130,6 +132,7 @@ func NewCmdUp(name, fullName string, f *osclientcmd.Factory, out io.Writer) *cob
 	cmd.Flags().IntVar(&config.ServerLogLevel, "server-loglevel", 0, "Log level for OpenShift server")
 	cmd.Flags().StringSliceVarP(&config.Environment, "env", "e", config.Environment, "Specify key value pairs of environment variables to set on OpenShift container")
 	cmd.Flags().BoolVar(&config.ShouldInstallMetrics, "metrics", false, "Install metrics (experimental)")
+	cmd.Flags().BoolVar(&config.AlwaysPull, "always-pull", false, "Always pull the latest images")
```
Comment (bparees) [inline pkg/bootstrap/docker/up.go]: so this affects infrastructure images as well as builder images and pod images?

Comment (bparees) [inline pkg/bootstrap/docker/up.go]: how do i tell it to use my local images (which may not match to "config.ImageVersion")?  specify version as "latest" and always-pull to false?

Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: > so this affects infrastructure images as well as builder images and pod images?

yes



> how do i tell it to use my local images (which may not match to "config.ImageVersion")? specify version as "latest" and always-pull to false?

yes (basically --version=latest) --always-pull will default to false.

Comment (bparees) [inline pkg/bootstrap/docker/up.go]: ok, i think always pull should be clearer about what it applies to  (or maybe that just belongs in the oc cluster up docs)

**Comment by bparees** [comment]
doc concerns aside, lgtm

File: pkg/bootstrap/docker/up.go (hunk 1 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: I agree, will update the wording here and update the cluster up doc.

**Comment by csrwng** [comment]
updated description, var names, and doc

File: contrib/completions/bash/openshift [hunk 2]
```diff
@@ -4904,6 +4904,7 @@ _openshift_cli_cluster_up()
     flags_with_completion=()
     flags_completion=()
 
+    flags+=("--always-pull")
```
Comment (bparees) [inline contrib/completions/bash/openshift]: need to regen completions/doc.

**Comment by bparees** [comment]
lgtm but i think there was still some question as to whether @smarterclayton was ok with the force pull behavior/overhead.  Since it's not the default, hopefully he is.

**Comment by smarterclayton** [comment]
--always-pull doesn't sound right as a flag name based on your description.  There is no "always" if this only happens once.

**Comment by csrwng** [comment]
It doesn't happen only once. If you specify the flag (--force-pull), you'd pull every time for every image that you use.

**Comment by csrwng** [comment]
@smarterclayton bump

**Comment by csrwng** [comment]
Closing in favor of 2 smaller PRs
one to set the default version, and another one to add a --force-pull flag if desired

**Pull request closed without merging**